programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -1245,21 +1548,88 @@ fn update_caller_account(
             return Err(Box::new(InstructionError::AccountDataTooSmall));
         }
         to_slice.copy_from_slice(from_slice);
-    } else if !is_loader_deprecated && realloc_bytes_used > 0 {
-        let to_slice = caller_account
-            .serialized_data
-            .get_mut(0..realloc_bytes_used)
-            .ok_or(SyscallError::InvalidLength)?;
+    } else if realloc_bytes_used > 0 {
+        // In the is_loader_deprecated case, we must have failed with
+        // InvalidRealloc by now.
+        debug_assert!(!is_loader_deprecated);
+
+        let to_slice = {
+            // If a callee reallocs an account, we write into the caller's
+            // realloc region regardless of whether the caller has write
+            // permissions to the account or not. If the callee has been able to
+            // make changes, it means they had permissions to do so, and here
+            // we're just going to reflect those changes to the caller's frame.
+            //
+            // Therefore we temporarily configure the realloc region as writable
+            // then set it back to whatever state it had.
+            let realloc_region = caller_account
+                .realloc_region(memory_mapping, is_loader_deprecated)?
+                .unwrap(); // unwrapping here is fine, we asserted !is_loader_deprecated
+            let original_state = realloc_region.state.replace(MemoryState::Writable);
+            defer! {
+                realloc_region.state.set(original_state);
+            };
+
+            translate_slice_mut::<u8>(
+                memory_mapping,
+                caller_account
+                    .vm_data_addr
+                    .saturating_add(caller_account.original_data_len as u64),
+                realloc_bytes_used as u64,
+                invoke_context.get_check_aligned(),
+                invoke_context.get_check_size(),
+            )?
+        };
         let from_slice = callee_account
             .get_data()
             .get(caller_account.original_data_len..post_len)
             .ok_or(SyscallError::InvalidLength)?;
+        if to_slice.len() != from_slice.len() {
+            return Err(Box::new(InstructionError::AccountDataTooSmall));
+        }
```

Comment:
nit: i wonder how this error condition could be triggered, sans the integer math saturation (which can't happen in this context)....

to_slice.len() actually is `realloc_bytes_used`, which in turn is `post_len.saturating_sub(caller_account.original_data_len)`, and `from_slice.len()` is `caller_account.original_data_len..post_len`

maybe better off promoting to `assert!()` or demoting to `debug_assert!()`? (hint: I'm pro assert guy)

Reply:
Unless $future_bugs, this condition can't happen, in the same way that the same check in the !direct_mapping path can't happen. It's a "just in case" defensive check, which as I mentioned I'm not in favor of having, but it doesn't seem like changing our policy belongs to this PR?